packages/lesswrong/server/callbacks/rateLimits.ts

```diff
@@ -78,29 +80,114 @@ async function enforcePostRateLimit (user: DbUser) {
 
 }
 
-async function enforceCommentRateLimit(user: DbUser) {
+async function enforceCommentRateLimit(user: DbUser, comment: DbComment) {
+  const rateLimit = await rateLimitDateWhenUserNextAbleToComment(user);
+  if (rateLimit) {
+    const {nextEligible, rateLimitType:_} = rateLimit;
+    if (nextEligible > new Date()) {
+      throw new Error(`Rate limit: You cannot comment until ${nextEligible}`);
+    }
+  }
+  
+  if (comment.postId) {
+    const postSpecificRateLimit = await rateLimitGetPostSpecificCommentLimit(user, comment.postId);
+    if (postSpecificRateLimit) {
+      const {nextEligible, rateLimitType:_} = postSpecificRateLimit;
+      if (nextEligible > new Date()) {
+        throw new Error(`Rate limit: You cannot comment on this post until ${nextEligible}`);
+      }
+    }
+  }
+}
+
+type RateLimitReason = "moderator"|"lowKarma"|"universal"
+
+/**
+ * If the user is rate-limited, return the date/time they will next be able to
+ * comment. If they can comment now, returns null.
+ */
+export async function rateLimitDateWhenUserNextAbleToComment(user: DbUser): Promise<{
+  nextEligible: Date,
+  rateLimitType: RateLimitReason
+}|null> {
   if (userIsAdmin(user) || userIsMemberOf(user, "sunshineRegiment")) {
-    return;
+    return null;
   }
 
+  // If moderators have imposed a rate limit on this user, enforce that
   const moderatorRateLimit = await getModeratorRateLimit(user)
   if (moderatorRateLimit) {
     const hours = getTimeframeForRateLimit(moderatorRateLimit.type)
 
-    const commentsInPastTimeframe = await userNumberOfItemsInPastTimeframe(user, Comments, hours)
-  
-    if (commentsInPastTimeframe > 0) {
-      throw new Error(MODERATOR_ACTION_TYPES[moderatorRateLimit.type]);
+    const mostRecentInTimeframe = await getNthMostRecentItemDate({
+      user, collection: Comments,
+      n: 1,
+      cutoffHours: hours,
+    });
+    if (mostRecentInTimeframe) {
+      return {
+        nextEligible: moment(mostRecentInTimeframe).add(hours, 'hours').toDate(),
+        rateLimitType: "moderator",
+      }
     }
-  
   }
+  
+  // commented out for now until EA Forum reviews
+  // If less than 30 karma, you are also limited to no more than 3 comments per
+  // 0.5 hours.
+  // if (!(user.karma >= 30)) {
+  //   const thirdMostRecentCommentDate = await getNthMostRecentItemDate({
+  //     user, collection: Comments,
+  //     n: 3,
+  //     cutoffHours: 0.5,
+  //   });
```

Comment:
Actually we decided to go with 4 comments instead of 3 for the first version, could you change that here?